### PR TITLE
en-GB and en-US locales now use 12 hour format

### PR DESCRIPTION
**Pull request by harogaston**
Original "HH:mm" (23:59) -> Now "hh:mm aa" (11:59 p.m.)

Before submitting a pull request, please make sure you have at least performed the following:

 - [X] read and followed the [CONTRIBUTING.md](https://github.com/edcarroll/ng2-semantic-ui/blob/master/CONTRIBUTING.md#) guide.
 - [X] linted, built and tested the changes locally.
 - [X] added/updated any applicable API documentation.
 - [X] added/updated any applicable demos.

### CODE REVIEW

**Comment by edcarroll** [comment]
Just want to hold off on this one until the 12:00 display is sorted for consistency, hope that's ok :)

**Comment by harogaston** [comment]
Sorry, I did not get what you mean. Nevertheless I'm sure you are right. But I'm curious.

**Comment by edcarroll** [comment]
Sorry for the delay on getting an answer to what I mean!

So the issue with using `aa` and its siblings is the fact that they currently aren't correctly mapped to `date-fns`. If you look at the code that converts the locale values `months`, `monthsShort` e.t.c, it doesn't have any logic for converting time values.

Can you please add this in? Take a look at the [date-fns locales](https://github.com/date-fns/date-fns/tree/master/src/locale/en-US) for an example of how they encode the values

**Comment by harogaston** [comment]
I'm sorry but I don't get your point. Can you give me an example of what doesn't work?

date-fns locales have this:
```
  var timeOfDayValues = {
    uppercase: ['AM', 'PM'],
    lowercase: ['am', 'pm'],
    long: ['a.m.', 'p.m.']
  }

  var localize = {
    ordinalNumber: ordinalNumber,
    weekday: buildLocalizeFn(weekdayValues, 'long'),
    weekdays: buildLocalizeArrayFn(weekdayValues, 'long'),
    month: buildLocalizeFn(monthValues, 'long'),
    months: buildLocalizeArrayFn(monthValues, 'long'),
    timeOfDay: buildLocalizeFn(timeOfDayValues, 'long', function (hours) {
      return (hours / 12) >= 1 ? 1 : 0
    }),
    timesOfDay: buildLocalizeArrayFn(timeOfDayValues, 'long')
  }
```

Is that `timeOfDay` what you are talking about?

**Comment by edcarroll** [comment]
Exactly - at the moment, `ng2-semantic-ui` takes the locale values, transforms them with custom versions of those functions (`buildLocalizeFn` , e.t.c.) and outputs a locale that `date-fns` understands.

At the moment, that doesn't include `timeOfDayValues` - so if someone wants to use it in a country that doesn't use `PM`, but instead `GI` (for the sake of this example), they can't currently just use `aa`.

Does that make sense?

**Comment by harogaston** [comment]
Oh I see your point now. Sorry it took me a while. I'll try something during the weekend. Sorry about the repeated request for clarification.

**Comment by edcarroll** [comment]
No worries 😄 should be a quick one I hope, as I think you'll be able to reuse the functions in `datepicker/helpers/date-fns.ts`.

**Comment by edcarroll** [comment]
Ah also - when you've put a solution in place, please ping the various locale creators so they know to update their translations with the correct am-pm values (or steal them from `date-fns` 😉)

**Comment by harogaston** [comment]
@edcarroll Do you what to have all the alternatives that `date-fns` has [uppercase|lowercase|long] or just one that maps to all three?

```
var timeOfDayValues = {
  uppercase: ['AM', 'PM'],
  lowercase: ['am', 'pm'],
  long: ['a.m.', 'p.m.']
};
```

**Comment by edcarroll** [comment]
@harogaston let's go for all three, I feel clarity is better here

File: src/behaviors/localization/locales/en-US.ts [hunk 1]
```diff
@@ -4,8 +4,8 @@ const enUS:IPartialLocaleValues = {
     datepicker: {
         firstDayOfWeek: 0,
         formats: {
-            time: "HH:mm",
-            datetime: "MMMM D, YYYY HH:mm",
+            time: "hh:mm aa",
```
Comment (edcarroll) [inline src/behaviors/localization/locales/en-US.ts]: Does this invalidate the need for the other PR?

File: src/modules/datepicker/helpers/date-fns.ts [hunk 2]
```diff
@@ -13,20 +13,28 @@ interface IDateFnsCustomLocale {
...
     };
     options?:{
         weekStartsOn?:number;
     };
 }
 
-function buildLocalizeFn(values:IDateFnsLocaleValues, defaultType:string):DateFnsHelper<number, string> {
-    return (dirtyIndex, { type } = { type: defaultType }) => values[type][dirtyIndex];
+function buildLocalizeFn(values:IDateFnsLocaleValues, defaultType:string,
+                         indexCallback?:(oldIndex:number) => number):DateFnsHelper<number, string> {
```
Comment (edcarroll) [inline src/modules/datepicker/helpers/date-fns.ts]: Could you put each arg on its own line please? Leave a space before the 1st statement in the function also when breaking like this

File: src/behaviors/localization/locales/en-US.ts (hunk 1 above)
Comment (edcarroll) [inline src/behaviors/localization/locales/en-US.ts]: Wait I'm an idiot ignore me

**Comment by edcarroll** [comment]
Final thing regarding all of your commits, please make sure you are following the commit message style in the contribution guidelines - i.e. feat(datepicker): Added support for 12 hour locales

**Comment by harogaston** [comment]
For better results it should be tested against [feat(datepicker) Popup now honors locale and pickerLocaleOverrides](https://github.com/edcarroll/ng2-semantic-ui/pull/215) but it works independently as well.

For now only two values for `timeOfDay` are supported, something like ["in the morning", "in the afternoon", "in the evening"] could be added later. it won't work just now because I think `localization.service.ts:deepClone()` and/or `localization.service.ts:deepExtend` do not work well with function objects which will be needed in locale definitions.

File: src/modules/datepicker/helpers/date-fns.ts (hunk 2 above)
Comment (harogaston) [inline src/modules/datepicker/helpers/date-fns.ts]: Working on it

**Comment by edcarroll** [comment]
@harogaston have just merged that PR, do you want to pull master into this branch and do a quick test?

No worries re only 2 values - does `date-fns` support more than 2?

**Comment by harogaston** [comment]
> Final thing regarding all of your commits, please make sure you are following the commit message style in the contribution guidelines - i.e. feat(datepicker): Added support for 12 hour locales

Sorry about that. I also have a few other similar questions. I'll talk to you via gitter later.

**Comment by edcarroll** [comment]
Don't forgot to mark the review comments as done to close them 😄

**Comment by harogaston** [comment]
> @harogaston have just merged that PR, do you want to pull master into this branch and do a quick test?

I have tested it including those changes already.

> No worries re only 2 values - does date-fns support more than 2?

Yes they allow things like:
```
var timeOfDayValues = {
  uppercase: ['AM', 'PM'],
  lowercase: ['am', 'pm'],
  long: ['du matin', 'de l’après-midi', 'du soir']
};
```

But for that to work locales will have to provide their own parsing functions.

**Comment by harogaston** [comment]
> Don't forgot to mark the review comments as done to close them 😄

I don't know how to...

**Comment by edcarroll** [comment]
Ahh ok let's leave the further customisation as a future feature 😄

**Pull request merged**